Baikal/PostEffects/ML/denoiser.cpp

```diff
@@ -22,5 +52,142 @@ namespace Baikal
                                                    height,
                                                    input_channels);
         }
+
+        MLDenoiser::MLDenoiser(const CLWContext& context, Inference::Ptr inference, MLDenoiserInputs inputs)
+                   : m_inference(std::move(inference))
+        {
+            m_context = std::make_unique<CLWContext>(context);
+            m_primitives = std::make_unique<CLWParallelPrimitives>(context);
+
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+            auto channels_num = std::get<2>(shape);
+
+            size_t elems_count = sizeof(Tensor::ValueType) * width * height * channels_num;
+
+            m_device_cache = std::make_unique<CLWBuffer<char>>(
+                CLWBuffer<char>::Create(*m_context, CL_MEM_READ_WRITE, elems_count));
+
+            m_host_cache = std::make_unique<std::uint8_t[]>(elems_count);
+
+            // compute memory layout
+            switch (inputs)
+            {
+                case MLDenoiserInputs::kColorDepthNormalGloss7:
+                {
+                    m_layout.emplace(OutputType::kColor, 3 * width * height);
+                    m_layout.emplace(OutputType::kDepth, width * height);
+                    m_layout.emplace(OutputType::kViewShadingNormal, 2 * width * height);
+                    m_layout.emplace(OutputType::kGloss, width * height);
+                }
+            }
+        }
+
+        template <class ClType, class Type>
+        void MLDenoiser::ProcessOutput(const CLWBuffer<RadeonRays::float3>& input,
+                                       Tensor::ValueType* host_mem)
+        {
+            auto input_buf = CLWBuffer<ClType>::CreateFromClBuffer(input);
+            auto normalized_buf = CLWBuffer<ClType>::CreateFromClBuffer(*m_device_cache);
+
+            m_primitives->Normalize(0,
+                                    input_buf,
+                                    normalized_buf,
+                                    (int)input_buf.GetElementCount()).Wait();
+
+            m_context->ReadBuffer<Type>(0,
+                                        CLWBuffer<Type>::CreateFromClBuffer(
+                                        normalized_buf),
+                                        (Type*)(host_mem),
+                                        input_buf.GetElementCount()).Wait();
+        }
+
+        void MLDenoiser::Apply(InputSet const& input_set, Output& output)
+        {
+            auto shape = m_inference->GetInputShape();
+            auto width = std::get<0>(shape);
+            auto height = std::get<1>(shape);
+
+            auto tensor = m_inference->GetInputTensor();
+            auto host_mem = tensor.data();
+
+            for (const auto& input : input_set)
```

Comment:
This code depends on the inputs iteration order which in the case of std::map is the `Renderer::OutputType`  values order. Instead we should loop through m_layout (m_layout should be something like `std::vector<std::pair<Renderer::OutputType, size_t>>`) the and get corresponding inputs with `input_set.at(type)`.